### DISCUSSION TITLE
Configuring subdomain in Laragon

### DISCUSSION THREAD

**Discussion by a-smily** [leokhoa/laragon#584] [Q&A] [answered]
Hello, 
I want to create a new laravel project with subdomains and I got problems to set the environment for the localhost.

The domain-structure should be:
mydomain.test
admin.mydomain.test
app.mydomain.test

I tried everything I found with Google, but nothing works.
It would be nice to have a howto or a working example to achieve this?
I tried quiet everything I found by Google, but nothing works (for example [this,](https://stackoverflow.com/questions/71053265/configuring-subdomain-in-laragon))

Thank you.

**Comment by Pen-y-Fan** [comment, marked as answer]
The explanation you have found is correct, I can give more information on how Laragon works and add more detail.

When a new project is created in **c:\laragon\www** Laragon will automatically detect the folder, it will:

- add an entry in the host in C:\Windows\System32\drivers\etc\host
- create an Apache conf file in C:\laragon\etc\apache2\sites-enabled

Each time a project is added to the www directory the **auto** file will be re-created, it is not advisable to edit them. 

Example:

I assume each of the projects has its own entry point, therefore create the three projects in the www directory:

admin-project
app-project
main-project

Laragon will create three entries in the host file:

```txt
127.0.0.1      admin-project.test       #laragon magic! 
127.0.0.1      app-project.test       #laragon magic! 
127.0.0.1      main-project.test       #laragon magic! 
```

Laragon will also create three Apache conf files:

```txt
auto.admin-project.test.conf
auto.app-project.test.conf
auto.main-project.test.conf
```

Copy the three #laragon magic!  lines in the host file and amended them to the required subdomain

```text
127.0.0.1      admin.myproject.local
127.0.0.1      app.myproject.local
127.0.0.1      myproject.local
```

Tip: Note the domain change from .test to .local, this is to separate the Laragon domain from manually configured ones. Feel free to amend Laragon to use .local and .test for manual.

Next, navigate to the **sites-enabled** folder. Copy each of the **auto** Apache conf files and rename them:

```txt
manual.admin.myproject.local.conf
manual.app.myproject.local.conf
manual.myproject.local.conf
```

Edit each of the three manual files and change the **define SITE** e.g. the **app** site will be:

```txt
define SITE "app.myproject.local"
```

For the **main** site also remove the wildcard: ServerAlias *.${SITE}, as this may cause problems with the subdomains

Restart Apache and you can now navigate each site by its subdomain.

- https://myproject.local
- https://admin.myproject.local
- https://app.myproject.local

Note: Each site must have an index.html or index.php file in the project folder or public subfolder for Laragon Magic to work, if you have another entry point the Apache file may need to be amended. For example, How to develop a Node.js app with Laragon ? #455

Tip: if you haven't enabled SSL in Laragon see my blog: [How to set up Laragon on a new Windows computer (part 4) - Enable security](https://pen-y-fan.github.io/2022/11/10/how-to-set-up-laragon-on-a-new-windows-computer-part-4/)

I hope this helps, if you have any further questions please ask.

**Comment by a-smily** [comment]
Thanks very much for your fast answer, your explanation works fine.
The trick was to create the manual config files for each domain/subdomain.

So for the sake of completeness, the content of my manuals:


For the main domain ->**manual.myproject.test.conf**:

> ```
> define ROOT "D:/laragon/www/test/livewire3/livewire3-filament3/public/"
> define SITE "myproject.test"
> 
> <VirtualHost *:80> 
>     DocumentRoot ${ROOT}
>     ServerName ${SITE}
>     ServerAlias ${SITE}
>     <Directory "${ROOT}">
>         AllowOverride All
>         Require all granted
>     </Directory>
> </VirtualHost>
> ```


And for the "admin" Subdomain ->**manual.admin.myproject.test.conf**:

> ```
> define ROOT "D:/laragon/www/test/livewire3/livewire3-filament3/public/"
> define SITE "admin.myproject.test"
> 
> <VirtualHost *:80> 
>     DocumentRoot ${ROOT}
>     ServerName ${SITE}
>     ServerAlias *.${SITE}
>     <Directory "${ROOT}">
>         AllowOverride All
>         Require all granted
>     </Directory>
> </VirtualHost>
> ```

**Comment by gul-ahmad** [comment]
but I think if we have some thing like multitenant application where subdomain created on the fly we don't need multiple projects for subdomain I mean our main application   abc.test and the subdomain is user1.abc.test and then user2.abc.test and so on...so the above discussion does not look or support this case as i have tried it practically any input or solution for this is welcome

**Reply by RachidMA** [reply]
i have an issue to configure laragon for multi tenant app. main domain working fine but subdomains, tenant-1.myapp.test or tenant-2.myapp.test are not been served. do you have any idea what could be the solution or if you have some more resources to look at. thank you

**Reply by gul-ahmad** [reply]
I resolved it ,and its an old story now. But I can try to figure it for u. Lets connect

**Reply by RachidMA** [reply]
Thank you very much. 
I am working localy on laravel multi tenancy. Using laragon I have tried to configure my hosts file. And it's seems that I need to enter each subdomain created to each tenant manually.

**Reply by RachidMA** [reply]
Is that correct? Or is there an automatic way to do so. Even setting wildcard in my hosts file and auto.myapp.test.conf file didn't work

**Reply by gul-ahmad** [reply]
Yes i just checked my project which is as well a multi tenancy project ,We have to add each subdomain manually created for tenant like e.g tenant1.learning.test ,tenant2.learning.test etc in hosts file and so on ,as per my findings there is no automatic way for it.

**Reply by RachidMA** [reply]
So painful.  It maybe a way to create window script to add the subdomain when new tenant is created. At least during development. I am working on it.  Thank you anyway

**Reply by gul-ahmad** [reply]
Its my pleasure and you are welcome .You can watch or learn more about different topics at my channel as well regarding laravel 

**https://www.youtube.com/channel/UCYpHilpRt59iumPI1sEPh2w**

**Reply by RachidMA** [reply]
Sure. I will save the link
